internal/alb/ls/listener.go

```diff
@@ -250,3 +264,76 @@ func (controller *defaultController) buildDefaultActions(ctx context.Context, op
 	}
 	return buildActions(ctx, authCfg, options.IngressAnnos, backend, options.TGGroup)
 }
+
+// inferCertARNs retrieves a set of certificates from ACM that matches the ingress' hosts list
+func inferCertARNs(acmsvc aws.ACMAPI, ingress *extensions.Ingress, logger *log.Logger) ([]string, error) {
+	var certArns []string
+	var seen = map[string]bool{}
+	var ingressHosts = uniqueHosts(ingress)
+
+	certs, err := acmsvc.ListCertificates([]string{acm.CertificateStatusIssued})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range certs {
+		for _, h := range ingressHosts {
+			if domainMatchesHost(aws.StringValue(c.DomainName), h) {
+				logger.Infof("Domain name '%s', matches TLS host '%v', adding to Listener", aws.StringValue(c.DomainName), h)
+				if !seen[aws.StringValue(c.CertificateArn)] {
+					certArns = append(certArns, aws.StringValue(c.CertificateArn))
+					seen[aws.StringValue(c.CertificateArn)] = true
+				}
+			} else {
+				logger.Debugf("Ignoring domain name '%s', doesn't match '%s'", aws.StringValue(c.DomainName), h)
+			}
+		}
+	}
+
+	return certArns, nil
+}
+
+func domainMatchesHost(domainName string, tlsHost string) bool {
+	if strings.HasPrefix(domainName, "*.") {
+		ds := strings.Split(domainName, ".")
+		hs := strings.Split(tlsHost, ".")
+
+		if len(ds) != len(hs) {
+			return false
+		}
+
+		for i, dp := range ds {
+			if i == 0 && dp == "*" {
+				continue
+			}
+			if dp != hs[i] {
+				return false
+			}
+		}
+		return true
+	}
+
+	return domainName == tlsHost
+}
+
+func uniqueHosts(ingress *extensions.Ingress) []string {
+	var result []string
```

Comment:
I think all these `set` operations via `seen` trick can be simplified via sets.NewString()

Reply:
Gotcha, fixed.